File: src/leetCode/java/Solution787.java
package leetCode.java;

import java.util.*;

//Shortest Path Bellman–Ford Algorithm O(kE) O(V), E = flights.length, V = n
class Solution787 {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        //Bellman–Ford可以在負權環
        int[] prices = new int[n];
        Arrays.fill(prices, Integer.MAX_VALUE);

        prices[src] = 0;
        for(int i=0; i < k + 1; i++){//Arrays.copyOf(陣列, 長度)
            int[] newPrices = Arrays.copyOf(prices, prices.length);//記得要複製用來更新
            for(int[] flight : flights){
                int a = flight[0];
                int b = flight[1];
                int cost = flight[2];
                if(prices[a] < Integer.MAX_VALUE)
                    newPrices[b] = Math.min(newPrices[b], prices[a] + cost);
            }
            prices = newPrices;
        }
        return prices[dst] == Integer.MAX_VALUE ? -1 : prices[dst];
    }
}
/* n = 4, flights = [[0,1,100],[1,2,100],[2,0,100],[1,3,600],[2,3,200]], src = 0, dst = 3, k = 1
          0   1   2   3     k
prices =  0  INF INF INF
          0  100 INF INF    0
		  0  100 200 700    1
		  0  100 200 400    2
*/

File: src/leetCode/c++/Solution787.cpp
//Shortest Path Bellman–Ford Algorithm O(kE) O(V), E = flights.size(), V = n
class Solution787 {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<int> dist(n, INT_MAX);
        dist[src] = 0;
        for(int i = 0; i < k + 1; ++i){//可轉機k次, 即可走k+1
            vector<int> temp = dist;//記得用temp避免混淆
            for(auto f : flights){
                if(temp[f[0]] != INT_MAX)
                    dist[f[1]] = min(dist[f[1]], temp[f[0]] + f[2]);
            }
        }
        return dist[dst] == INT_MAX ? -1 : dist[dst];
    }
};
/* n = 4, flights = [[0,1,100],[1,2,100],[2,0,100],[1,3,600],[2,3,200]], src = 0, dst = 3, k = 1
k       0   1   2   3
0       0
1       0   100
2       0   100 200 700
3       0   100 200 400

*/